interfaces/builtin/content.go

```diff
@@ -0,0 +1,160 @@
+	}
+
+	// check that we have either a read or write path
+	rpath, rok := slot.Attrs["read"].([]interface{})
+	wpath, wok := slot.Attrs["write"].([]interface{})
+	if !rok && !wok {
+		return fmt.Errorf("content must contain a read or write path")
+	}
+	if len(rpath) == 0 && len(wpath) == 0 {
+		return fmt.Errorf("read or write path must be set")
+	}
+
+	// go over both paths
+	paths := rpath
+	paths = append(paths, wpath...)
+	for _, p := range paths {
+		if strings.Contains(p.(string), "..") {
+			return fmt.Errorf("relative path not allowed")
+		}
+	}
+
+	return nil
+}
+
+func (iface *ContentInterface) SanitizePlug(plug *interfaces.Plug) error {
+	if iface.Name() != plug.Interface {
+		panic(fmt.Sprintf("plug is not of interface %q", iface))
+	}
+	target, ok := plug.Attrs["target"].(string)
+	if !ok || len(target) == 0 {
+		return fmt.Errorf("content plug must contain target path")
+	}
+	if strings.Contains(target, "..") {
+		return fmt.Errorf("relative path not allowed")
+	}
+
+	return nil
+}
+
+func (iface *ContentInterface) ConnectedSlotSnippet(plug *interfaces.Plug, slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	switch securitySystem {
+	case interfaces.SecurityAppArmor, interfaces.SecuritySecComp, interfaces.SecurityDBus, interfaces.SecurityUDev, interfaces.SecurityMount:
+		return nil, nil
+	default:
+		return nil, interfaces.ErrUnknownSecurity
+	}
+}
+
+func (iface *ContentInterface) PermanentSlotSnippet(slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+
+	switch securitySystem {
+	case interfaces.SecurityAppArmor, interfaces.SecuritySecComp, interfaces.SecurityDBus, interfaces.SecurityUDev, interfaces.SecurityMount:
+		return nil, nil
+	default:
+		return nil, interfaces.ErrUnknownSecurity
+	}
+}
+
+// path is an internal helper that extract the "read" and "write" attribute
+// of the slot
+func (iface *ContentInterface) path(slot *interfaces.Slot, name string) []string {
+	if name != "read" && name != "write" {
+		panic("internal error, path can only be used with read/write")
+	}
+
+	paths, ok := slot.Attrs[name].([]interface{})
+	if !ok {
+		return nil
+	}
+
+	out := make([]string, len(paths))
+	for i, p := range paths {
+		out[i], ok = p.(string)
+		if !ok {
+			return nil
+		}
+
+	}
+	return out
+}
+
+func mountEntry(plug *interfaces.Plug, slot *interfaces.Slot, relSrc string, mntOpts string) string {
+	dst := plug.Attrs["target"].(string)
+	dst = filepath.Join(plug.Snap.MountDir(), dst)
+	src := filepath.Join(slot.Snap.MountDir(), relSrc)
+	return fmt.Sprintf("%s %s none bind%s 0 0", src, dst, mntOpts)
+}
+
+func (iface *ContentInterface) ConnectedPlugSnippet(plug *interfaces.Plug, slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	contentSnippet := bytes.NewBuffer(nil)
+	for _, r := range iface.path(slot, "read") {
+		fmt.Fprintln(contentSnippet, mountEntry(plug, slot, r, ",ro"))
```

Comment:
Aha, thank you! I indeed misunderstood. Just for the record (we talked on tg already). I am fine extracting it but would prefer to wait until its actually used outside of the `content` interface.